Overhaul ecosystem overview (#75)
esp-rs/book: src/overview/using-the-standard-library.md

# Using the Standard Library (`std`)

> Espressif provides a C-based development framework called [esp-idf][esp-idf-github] which has support for all Espressif chips starting with the ESP32; note that this framework does _not_ support the ESP8266. See also Section [Support for Espressif Products][rust-esp-book-overview-index].

Reviewer: ```suggestion
Espressif provides a C-based development framework called [esp-idf][esp-idf-github] which has, or will have, support for all Espressif chips starting with the ESP32; note that this framework does _not_ support the ESP8266. See also [Support for Espressif Products][rust-esp-book-overview-index].
```

[rust-esp-book-overview-index]: index.md#support-for-espressif-products

> `esp-idf` in turn provides a [newlib][newlib-env] environment with enough functionality to build the Rust standard library (`std`) on top of it. This is the approach that is being taken to enable `std` support on ESP devices.

Reviewer: ```suggestion
`esp-idf` in turn, provides a [newlib][newlib-env] environment with enough functionality to build the Rust standard library (`std`) on top of it. This is the approach that is being taken to enable `std` support on ESP devices.
```

# Using the Standard Library (`std`)

Espressif provides a C-based development framework called [esp-idf][esp-idf-github] which has support for all Espressif chips starting with the ESP32; note that this framework does _not_ support the ESP8266. See also Section [Support for Espressif Products][rust-esp-book-std-vs-no-std-support].

[rust-esp-book-std-vs-no-std-support]: ./comparing-std-and-no_std.html#support-for-espressif-products

`esp-idf` in turn provides a [newlib][newlib-env] environment with enough functionality to build the Rust standard library (`std`) on top of it. This is the approach that is being taken to enable `std` support on ESP devices.

[esp-idf-github]: https://github.com/espressif/esp-idf

> ### Supported `std` Features

Reviewer: ```suggestion
## Supported `std` Features
```

Espressif products support a lot of `std` features that exist in [esp-idf][esp-idf-github], including threads, mutexes and other synchronization primitives, collections, random number generation, sockets.

> [esp-rs/esp-idf-svc](https://github.com/esp-rs/esp-idf-svc) adds extra support for services/modules not available in the standard library, such as Wi-Fi management, NVS (non-volatile storage), networking services, such as `httpd` and `ping`, etc.

Reviewer: I think we can remove this line as it does not add any value and just leave the description of `esp-idf-svc` of the table/list

> [esp-rs/esp-idf-svc](https://github.com/esp-rs/esp-idf-svc) adds extra support for services/modules not available in the standard library, such as Wi-Fi management, NVS (non-volatile storage), networking services, such as `httpd` and `ping`, etc.

Author: Deleted this line

### Relevant `esp-rs` crates

> [esp-idf-github]: https://github.com/espressif/esp-idf

Author: This file had it defined twice. Deleted this instance.

| Repository            | Description                                                                                                   |
| --------------------- | ------------------------------------------------------------------------------------------------------------- |
| [esp-rs/esp-idf-svc]  | An implementation of [embedded-svc] using `esp-idf` drivers.                                                  |
| [esp-rs/esp-idf-hal]  | An implementation of the `embedded-hal` and other traits using the `esp-idf` framework.                       |
| [esp-rs/esp-idf-sys]  | Rust bindings to the `esp-idf` development framework. Gives raw (`unsafe`) access to drivers, Wi-Fi and more. |
| [esp-rs/embedded-svc] | Abstraction traits for embedded services (`WiFi`, `Network`, `Httpd`, `Logging`, etc.)                       |

[newlib]: https://sourceware.org/newlib/
[embedded-svc]: https://github.com/esp-rs/embedded-svc
[esp-idf-svc]: https://github.com/esp-rs/esp-idf-svc

> The aforementioned crates have interdependencies, and this relationship can be seen below.

Author: If there are similar interdependencies between `no_std` crates, then, I believe, it will be very useful to have such a diagram. 

Can you please create one?

# Using the Standard Library (`std`)

Espressif provides a C-based development framework called [esp-idf][esp-idf-github] which has support for all Espressif chips starting with the ESP32; note that this framework does _not_ support the ESP8266. See also Section [Support for Espressif Products][rust-esp-book-std-vs-no-std-support].

[rust-esp-book-std-vs-no-std-support]: ./comparing-std-and-no_std.html#support-for-espressif-products

> `esp-idf` in turn provides a [newlib][newlib-env] environment with enough functionality to build the Rust standard library (`std`) on top of it. This is the approach that is being taken to enable `std` support on ESP devices.

Reviewer: We need to decide how to format esp-idf, as some times is `esp-idf` and others esp-idf

Author: Sometimes it is also ESP-IDF.

We will address this issue in our style guide and then will make respective corrections.

    esp-idf-svc --> esp-idf-sys & esp-idf-hal & embedded-svc
```

> [embedded-svc]: https://github.com/esp-rs/embedded-svc

Reviewer: ```suggestion
### When you might want to use the Standard Library (`std`)

- Rich functionality: If your embedded system requires lots of functionality like support for networking protocols, file I/O, or complex data structures, you will likely want to use hosted-environment approach because std libraries provide a wide range of functionality that can be used to build complex applications relatively quickly and efficiently
- Portability: The `std` crate provides a standardized set of APIs that can be used across different platforms and architectures, making it easier to write code that is portable and reusable.
- Rapid development: The `std` crate provides a rich set of functionality that can be used to build applications quickly and efficiently, without worrying about low-level details.

[embedded-svc]: https://github.com/esp-rs/embedded-svc
```

I think adding this section from @JurajSadel [blogpost](https://blog.espressif.com/rust-embedded-a-development-power-duo-db43dae21206) can be really useful

Reviewer: Definitely something to take into account, we could maybe add some clarifications.
